### PR TITLE
mutable state checksums part#2: persistence changes

### PR DESCRIPTION
**Pull request by venkat1109**
This patch contains the cassandra persistence changes to support mutable state checksums. This is a follow up to #2941

### CODE REVIEW

File: schema/cassandra/cadence/versioned/v0.24/manifest.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
+  "CurrVersion": "0.24",
+  "MinCompatibleVersion": "0.23",
```
Comment (wxing1292) [inline schema/cassandra/cadence/versioned/v0.24/manifest.json]: 0.24?

File: common/persistence/persistenceInterface.go [hunk 2]
```diff
@@ -24,6 +24,8 @@ import (
 	"fmt"
 	"time"
 
+	"github.com/uber/cadence/common/checksum"
```
Comment (wxing1292) [inline common/persistence/persistenceInterface.go]: make fmt?

File: common/persistence/cassandra/cassandraPersistenceUtil.go [hunk 3]
```diff
@@ -2500,6 +2524,24 @@ func createReplicationInfoMap(
 	return rInfoMap
 }
 
+func createChecksum(result map[string]interface{}) checksum.Checksum {
+	if len(result) == 0 {
+		return checksum.Checksum{}
+	}
+	csum := checksum.Checksum{}
```
Comment (wxing1292) [inline common/persistence/cassandra/cassandraPersistenceUtil.go]:
```go
csum := checksum.Checksum{}
if len(result) == 0 {
	return csum
}
```

File: common/persistence/persistenceInterface.go [hunk 4]
```diff
@@ -244,6 +246,7 @@ type (
 		SignalInfos         map[int64]*SignalInfo
 		SignalRequestedIDs  map[string]struct{}
 		BufferedEvents      []*DataBlob
+		Checksum            checksum.Checksum
 	}
```
Comment (wxing1292) [inline common/persistence/persistenceInterface.go]: i would put checksum in a seperate group, e.g.

```go
InternalWorkflowMutableState struct {
		ExecutionInfo    *InternalWorkflowExecutionInfo
		ReplicationState *ReplicationState
		VersionHistories *DataBlob

		ActivityInfos       map[int64]*InternalActivityInfo
		TimerInfos          map[string]*TimerInfo
		ChildExecutionInfos map[int64]*InternalChildExecutionInfo
		RequestCancelInfos  map[int64]*RequestCancelInfo
		SignalInfos         map[int64]*SignalInfo
		SignalRequestedIDs  map[string]struct{}
		BufferedEvents      []*DataBlob

		Checksum            checksum.Checksum
}
```

File: common/persistence/persistenceInterface.go [hunk 5]
```diff
@@ -380,6 +383,7 @@ type (
 		ReplicationTasks []Task
 
 		Condition int64
+		Checksum  checksum.Checksum
```
Comment (wxing1292) [inline common/persistence/persistenceInterface.go]: same as above

File: common/persistence/persistenceInterface.go [hunk 6]
```diff
@@ -402,6 +406,7 @@ type (
...
```
Comment (wxing1292) [inline common/persistence/persistenceInterface.go]: same as above

**Comment by wxing1292** [comment]
one more thing, where is the impl for SQL?

File: schema/cassandra/cadence/versioned/v0.24/manifest.json (hunk 1 above)
Comment (venkat1109) [reply schema/cassandra/cadence/versioned/v0.24/manifest.json]: This change is backwards compatible with 0.23 and so this is set to 0.23

File: common/persistence/persistenceInterface.go (hunk 4 above)
Comment (venkat1109) [reply common/persistence/persistenceInterface.go]: I assume this is simply personal preference. In general, this is something that is not auto enforcable by gofmt etc. And so, over time, your personal preference for grouping this separately is not sustainable. I will group this separately for now. But philosophically, I am not a big fan these type of semi-enforced groupings (unless there is strong reasons like struct packing in C/C++)

**Comment by venkat1109** [comment]
> one more thing, where is the impl for SQL?
SQL is currently not supported. I will consider adding support for it after rolling this out. There will be a github task created though for anyone to pick that up.

**Pull request merged**